oauth2client/contrib/django_util/__init__.py

```diff
@@ -238,75 +339,136 @@ def __init__(self, settings_instance):
 
 
 oauth2_settings = OAuth2Settings(django.conf.settings)
 
+_CREDENTIALS_KEY = 'google_oauth2_credentials'
+
+
+def get_storage(request):
+    """ Gets a Credentials storage object provided by the Django OAuth2 Helper
+    object.
+
+    Args:
+        request: Reference to the current request object.
+
+    Returns:
+       An :class:`oauth2.client.Storage` object.
+    """
+    storage_model = oauth2_settings.storage_model
+    user_property = oauth2_settings.storage_model_user_property
+    credentials_property = oauth2_settings.storage_model_credentials_property
+
+    if storage_model:
+        module_name, class_name = storage_model.rsplit('.', 1)
+        module = importlib.import_module(module_name)
+        storage_model_class = getattr(module, class_name)
+        return storage.DjangoORMStorage(storage_model_class,
+                                        user_property,
+                                        request.user,
+                                        credentials_property)
+    else:
+        # use session
+        return dictionary_storage.DictionaryStorage(
+            request.session, key=_CREDENTIALS_KEY)
+
 
 def _redirect_with_params(url_name, *args, **kwargs):
     """Helper method to create a redirect response that uses GET URL
-    parameters."""
+    parameters.
+
+    Args:
+        url_name: The name of the url to redirect to.
+        kwargs: the query string param and their values to build.
 
+    Returns:
+        A properly formatted redirect string.
+    """
     url = urlresolvers.reverse(url_name, args=args)
     params = parse.urlencode(kwargs, True)
     return "{0}?{1}".format(url, params)
 
 
+def _credentials_from_request(request):
+    """Gets the authorized credentials for this flow, if they exist."""
+    # ORM storage requires a logged in user
+    if (oauth2_settings.storage_model is None or
+            request.user.is_authenticated()):
+        return get_storage(request).get()
+    else:
+        return None
+
+
 class UserOAuth2(object):
     """Class to create oauth2 objects on Django request objects containing
     credentials and helper methods.
     """
 
     def __init__(self, request, scopes=None, return_url=None):
-        """Initialize the Oauth2 Object
-        :param request: Django request object
-        :param scopes: Scopes desired for this OAuth2 flow
-        :param return_url: URL to return to after authorization is complete
-        :return:
+        """Initialize the Oauth2 Object.
+
+        Args:
+            request: Django request object.
+            scopes: Scopes desired for this OAuth2 flow.
+            return_url: The url to return to after the OAuth flow is complete,
+                 defaults to the request's current URL path.
         """
         self.request = request
         self.return_url = return_url or request.get_full_path()
-        self.scopes = set(oauth2_settings.scopes)
+        self._scopes = set(oauth2_settings.scopes)
```

Comment:
When possible, avoid mutating name-value assignments, _especially in constructors_. Can this be
```
if scopes:
  self._scopes = set(oauth2_settings.scopes) | set(scopes)
else:
  self._scopes = set(oauth2_settings.scopes)
```
?